### PR TITLE
Adds domain correction method for invalid chars

### PR DESCRIPTION
**Pull request by ashraffouda**
Signed-off-by: Ashraf Fouda <ashraf.m.fouda@gmail.com>

### CODE REVIEW

File: jumpscale/sals/zos/gateway.py [hunk 1]
```diff
@@ -20,6 +22,17 @@ class GatewayGenerator:
 
     def __init__(self, explorer):
         self._gateways = explorer.gateway
+        
+    def correct_domain(self, domain):
```
Comment (zaibon) [inline jumpscale/sals/zos/gateway.py:26]: can you add a docstring explaining what this method does please. regex are always hard to read and deduct what they do.

Comment (ashraffouda) [reply jumpscale/sals/zos/gateway.py:26]: done

File: jumpscale/packages/tfgrid_solutions/chats/mattermost.py [hunk 2]
```diff
@@ -84,6 +84,7 @@ def domain_select(self):
 
         solution_name = self.solution_name.replace(".", "")
         full_domain = f"{self.threebot_name}-{solution_name}.{self.domain}"
+        self.domain = j.sals.zos.gateway.correct_domain(self.domain)
         while True:
             if j.tools.dnstool.is_free(full_domain):
```
Comment (xmonader) [inline jumpscale/packages/tfgrid_solutions/chats/mattermost.py:89]: shouldn't we validate here the cleanedup domain is free?

File: jumpscale/packages/tfgrid_solutions/chats/peertube.py [hunk 3]
```diff
@@ -149,6 +149,7 @@ def select_domain(self):
 
         self.gateway = domains[self.domain]
```
Comment (xmonader) [inline jumpscale/packages/tfgrid_solutions/chats/peertube.py:150]: would it affect to use self.domain here before correcting the domain?

File: jumpscale/packages/tfgrid_solutions/chats/threebot.py [hunk 4]
```diff
@@ -118,6 +118,7 @@ def domain_select(self):
         self.gateway = domains[self.domain]["gateway"]
         self.gateway_pool = domains[self.domain]["pool"]
         self.domain = f"{self.threebot_name}-{self.solution_name}.{self.domain}"
+        self.domain = j.sals.zos.gateway.correct_domain(self.domain)
```
Comment (xmonader) [inline jumpscale/packages/tfgrid_solutions/chats/threebot.py:121]: here also the self.domain was corrected but it was used before?

File: jumpscale/packages/tfgrid_solutions/chats/peertube.py (hunk 3 above)
Comment (ashraffouda) [reply jumpscale/packages/tfgrid_solutions/chats/peertube.py:150]: no, that is another thing, it first select domain, then and prefix to it 
self.doamin is like threefold.com 
self.domain after workds is user-solution.threefold.com
bad naming :)

File: jumpscale/packages/tfgrid_solutions/chats/threebot.py (hunk 4 above)
Comment (ashraffouda) [reply jumpscale/packages/tfgrid_solutions/chats/threebot.py:121]: same as above

File: jumpscale/packages/tfgrid_solutions/chats/mattermost.py (hunk 2 above)
Comment (ashraffouda) [reply jumpscale/packages/tfgrid_solutions/chats/mattermost.py:89]: handled

File: jumpscale/sals/zos/gateway.py [hunk 5]
```diff
@@ -37,7 +54,7 @@ def sub_domain(self, gateway_id: str, domain: str, ips: List[str], pool_id: int)
         for ip in ips:
             if not _is_valid_ip(ip):
                 raise Input(f"{ip} is not valid IP address")
-
+        self.correct_domain(domain)
```
Comment (xmonader) [inline jumpscale/sals/zos/gateway.py:57]: hmm what does this call do? as far as i understood it returns a corrected domain, but where is that captured?

**Review by xmonader — APPROVED** [review]

**Pull request merged**